api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,600 @@
+
+	resourceConsumers, resourceProviders, err := e.GetCurrentState(clusterName)
+	if err != nil {
+		return err
+	}
+
+	var incompatibilityError error
+	resourceProviders, unusedProviders, desiredScale, err := e.CalculateOptimizedState(clusterName, resourceConsumers, resourceProviders)
+
+	if err != nil {
+		if err, ok := err.(*errors.ServerError); ok && err.Code == errors.IncompatibleConsumerAndProvider {
+			incompatibilityError = err
+		} else {
+			return err
+		}
+	}
+
+	if err := e.ScaleToState(clusterName, desiredScale, unusedProviders); err != nil {
+		return err
+	}
+
+	return incompatibilityError
+}
+
+func (e *EnvironmentScaler) GetCurrentState(clusterName string) ([]scaler.ResourceConsumer, []*scaler.ResourceProvider, error) {
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	serviceConsumers, err := e.getServiceResourceConsumers(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	jobTaskConsumers, err := e.getJobTaskResourceConsumers(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resourceConsumers := append(serviceConsumers, jobTaskConsumers...)
+
+	return resourceConsumers, resourceProviders, nil
+}
+
+func (e *EnvironmentScaler) CalculateOptimizedState(clusterName string, resourceConsumers []scaler.ResourceConsumer, resourceProviders []*scaler.ResourceProvider) ([]*scaler.ResourceProvider, []*scaler.ResourceProvider, int, error) {
+	var incompatibilityErrs []error
+	var multiError error
+
+	// calculate for scaling up
+	resourceProviders, errs := e.calculateScaleUp(clusterName, resourceProviders, resourceConsumers)
+	for _, err := range errs {
+		// the IncompatibleConsumerAndProvider error is returned in situations where
+		// a resource consumer is too large to fit into an empty, brand-new resource
+		// provider; this means that the environment's instance size is too small
+		// for the consumer. in such a scenario, we want to perform scaling actions
+		// for other providers and consumers, but also log that there is a problem.
+		if err, ok := err.(*errors.ServerError); ok && err.Code == errors.IncompatibleConsumerAndProvider {
+			incompatibilityErrs = append(incompatibilityErrs, err)
+			continue
+		}
+
+		return nil, nil, 0, err
+	}
+
+	if len(incompatibilityErrs) > 0 {
+		multiError = errors.New(errors.IncompatibleConsumerAndProvider, errors.MultiError(incompatibilityErrs))
+	}
+
+	// calculate for scaling down
+	unusedProviders := e.calculateScaleDown(clusterName, resourceProviders)
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+
+	return resourceProviders, unusedProviders, desiredScale, multiError
+}
+
+func (e *EnvironmentScaler) ScaleToState(clusterName string, desiredScale int, unusedProviders []*scaler.ResourceProvider) error {
+	asg, err := readASG(e.Client.AutoScaling, clusterName)
+	if err != nil {
+		return err
+	}
+
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
```

Comment:
[nit] currentScale might be more appropriate in this context than currentCapacity

Reply:
✅